hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/supervisor/workload/weighted/WeightedWorkBalancer.java

```diff
@@ -0,0 +1,386 @@
+                .collect(toCollection(ArrayDeque::new));
+    }
+
+    private AssignmentPlan tryAssignUnassignedTasks(AssignmentPlan assignmentPlan) {
+        PriorityQueue<ConsumerTask> tasksToAssign = new PriorityQueue<>(HEAVIEST_TASK_FIRST);
+        tasksToAssign.addAll(assignmentPlan.getUnassignedTasks());
+        List<ConsumerTask> unassignedTasks = new ArrayList<>();
+        while (!tasksToAssign.isEmpty()) {
+            ConsumerTask consumerTask = tasksToAssign.poll();
+            Optional<ConsumerNode> candidate = selectConsumerNode(consumerTask, assignmentPlan.getConsumers());
+            if (candidate.isPresent()) {
+                candidate.get().assign(consumerTask);
+            } else {
+                unassignedTasks.add(consumerTask);
+            }
+        }
+        return new AssignmentPlan(unassignedTasks, assignmentPlan.getConsumers());
+    }
+
+    private Optional<ConsumerNode> selectConsumerNode(ConsumerTask consumerTask, Collection<ConsumerNode> consumers) {
+        return consumers.stream()
+                .filter(consumerNode -> consumerNode.isNotAssigned(consumerTask))
+                .filter(consumerNode -> !consumerNode.isFull())
+                .min(LIGHTEST_CONSUMER_FIRST);
+    }
+
+    private AssignmentPlan rebalance(AssignmentPlan plan) {
+        Collection<ConsumerNode> consumers = plan.getConsumers();
+        TargetConsumerLoad targetLoad = calculateTargetConsumerLoad(consumers);
+        List<ConsumerNode> overloadedConsumers = consumers.stream()
+                .filter(consumerNode -> isOverloaded(consumerNode, targetLoad))
+                .collect(toList());
+        for (ConsumerNode overloaded : overloadedConsumers) {
+            List<ConsumerNode> underloadedConsumers = consumers.stream()
+                    .filter(consumer -> !consumer.equals(overloaded))
+                    .filter(consumerNode -> isUnderloaded(consumerNode, targetLoad))
+                    .collect(toList());
+            for (ConsumerNode underloaded : underloadedConsumers) {
+                tryMoveOutTasksFromOverloaded(overloaded, underloaded, targetLoad);
+                trySwapTasks(overloaded, underloaded, targetLoad);
+                if (isBalanced(overloaded, targetLoad)) {
+                    break;
+                }
+            }
+        }
+        for (ConsumerNode consumerNode : consumers) {
+            logger.debug("Consumer {} weight after rebalance: {}", consumerNode, consumerNode.getWeight());
+        }
+        return new AssignmentPlan(plan.getUnassignedTasks(), consumers);
+    }
+
+    private TargetConsumerLoad calculateTargetConsumerLoad(Collection<ConsumerNode> consumers) {
+        int totalNumberOfTasks = consumers.stream().mapToInt(ConsumerNode::getAssignedTaskCount).sum();
+        int consumerCount = consumers.size();
+        int maxNumberOfTasksPerConsumer = consumerCount == 0 ? 0 : (totalNumberOfTasks + consumerCount - 1) / consumerCount;
```

Comment:
why `(totalNumberOfTasks + consumerCount - 1)` what is e.g `totalNumberOfTasks = 11` and `consumerCount = 3` ==> `(totalNumberOfTasks + consumerCount - 1) / consumerCount = (11 + 2) / 3 = 4.3` but maxNumberOfTasksPerConsumer will be `4` so effectively will have some tasks unassigned ?

Reply:
Good question! 

In the example you gave everything seems to work as expected: `maxNumberOfTasksPerConsumer` = `4` and we have `3` consumers, so there are `12` slots for tasks - we have one more slot than necessary. 

Anyway, I understand why you are confused here. What I wanted to achieve is simply rounding up. Java always rounds down when dividing two integers, but there is a formula that converts floors to ceilings (and I used it here): 
![image](https://user-images.githubusercontent.com/6481553/185803205-e4d119d2-af82-4ae6-9fb4-dae1919c1505.png)
source: [Floor and ceiling functions](https://en.wikipedia.org/wiki/Floor_and_ceiling_functions)

I can change this to something more intuitive:
```java
(totalNumberOfTasks / consumerCount) + (totalNumberOfTasks % consumerCount > 0 ? 1 : 0)
```

WDYT?

Reply:
cool, it would be easier to understand while reading :)

Reply:
done